src/dtls_connection_api.c

```diff
@@ -195,6 +195,74 @@ static ssize_t dtls_connection_net_send(gnutls_transport_ptr_t context, const vo
     return sendto(conn->sock, data, size, 0, (struct sockaddr *)&conn->addr, conn->addr_size);
 }
 
+static ssize_t dtls_connection_net_recv(gnutls_transport_ptr_t context, void *data, size_t size)
+{
+    device_connection_t *conn = (device_connection_t *)context;
+    int ret;
+    struct sockaddr addr;
+    socklen_t addr_size;
+
+    addr_size = sizeof(addr);
+    ret = recvfrom(conn->sock, data, size, 0, &addr, &addr_size);
+    if (ret < 0)
+    {
+        return ret;
+    }
+
+    if (addr_size == conn->addr_size
+        && memcmp(&addr, &conn->addr, sizeof(addr)) == 0)
+    {
+        return ret;
+    }
+
+    gnutls_transport_set_errno(conn->session, EAGAIN);
+
+    return -1;
+}
+
+static int dtls_connection_net_recv_timeout(gnutls_transport_ptr_t context, unsigned int ms)
+{
+    device_connection_t *conn = (device_connection_t *)context;
+    struct timeval tv;
+    fd_set fd;
+    int ret;
+    uint8_t buffer[1024];
+    size_t buffer_length;
+    struct sockaddr addr;
+    socklen_t addr_size;
+
+    tv.tv_sec = ms / 1000;
+    tv.tv_usec = (ms % 1000) * 1000;
+
+    FD_ZERO(&fd);
+    FD_SET(conn->sock, &fd);
+
+    ret = select(conn->sock + 1, &fd, NULL, NULL, &tv);
+    if (ret <= 0)
+    {
+        return ret;
+    }
+
+    buffer_length = sizeof(buffer);
+    addr_size = sizeof(addr);
+
+    ret = recvfrom(conn->sock, buffer, buffer_length, MSG_PEEK, &addr, &addr_size);
```

Comment:
Isn't this the same as `dtls_connection_net_recv`? If it is, it could be replaced with `return dtls_connection_net_recv(...);`